frontend/src/components/widgets/CustomComponent/ComponentInstance.test.tsx

```diff
@@ -0,0 +1,287 @@
+import { logWarning } from "lib/log"
+import { buildHttpUri } from "lib/UriUtil"
+
+// Mock log functions.
+jest.mock("lib/log")
+
+// Mock uri utils.
+jest.mock("lib/UriUtil")
+const mockedBuildHttpUri = buildHttpUri as jest.Mock
+mockedBuildHttpUri.mockImplementation(() => "registry/url")
+
+const getProps = (elementProps = {}): Props => ({
+  element: fromJS({
+    argsDataframe: [],
+    argsJson: '{"foo": "bar"}',
+    componentName: "some.component",
+    id: "some_id",
+    url: "some/url",
+    ...elementProps,
+  }),
+  registry: new ComponentRegistry(() =>
+    fromJS({
+      host: "streamlit.mock",
+      port: 80,
+      basePath: "",
+    })
+  ),
+  width: 100,
+  disabled: false,
+  widgetMgr: new WidgetStateManager(jest.fn()),
+})
+
+describe("ComponentInstance", () => {
+  let wrapper: ReactWrapper<Props, State, ComponentInstance>
+  let instance: ComponentInstance
+
+  beforeEach(() => {
+    const props = getProps()
+    const mountNode = document.createElement("div")
+    document.body.appendChild(mountNode)
+
+    // We need to mount the ComponentInstance to an existing DOM element.
+    // Otherwise, iframe contentWindow is not available.
+    wrapper = mount(<ComponentInstance {...props} />, { attachTo: mountNode })
+    instance = wrapper.instance()
+  })
+
+  describe("handleSetComponentValue", () => {
+    let handleSetComponentValue: (data: any, source: Source) => void
+    const source = { fromUi: true }
+
+    beforeEach(() => {
+      handleSetComponentValue = instance["handleSetComponentValue"]
+      WidgetStateManager.prototype.setJsonValue = jest.fn()
+    })
+
+    it("should warn that the `value` prop is missing", () => {
+      handleSetComponentValue({ value: undefined }, source)
+      expect(logWarning).toHaveBeenCalledWith(
+        `handleSetComponentValue: missing 'value' prop`
+      )
+      expect(WidgetStateManager.prototype.setJsonValue).not.toHaveBeenCalled()
+    })
+
+    it("should set widget value", () => {
+      handleSetComponentValue({ value: 1 }, source)
+      expect(WidgetStateManager.prototype.setJsonValue).toHaveBeenCalledWith(
+        "some_id",
+        1,
+        source
+      )
+    })
+  })
+
+  describe("handleSetFrameHeight", () => {
+    let handleSetFrameHeight: (data: any) => void
+
+    beforeEach(() => {
+      handleSetFrameHeight = instance["handleSetFrameHeight"]
+    })
+
+    it("should warn that the `height` prop is missing", () => {
+      handleSetFrameHeight({ height: undefined })
+      expect(logWarning).toHaveBeenCalledWith(
+        `handleSetFrameHeight: missing 'height' prop`
+      )
+      expect(instance["frameHeight"]).toEqual(0)
+    })
+
+    it("should set iframe height", () => {
+      handleSetFrameHeight({ height: 100 })
+      expect(instance["frameHeight"]).toEqual(100)
+    })
+
+    it("should warn that the `iframeRef` prop is missing", () => {
+      // @ts-ignore
+      instance["iframeRef"]["current"] = null
```

Comment:
Ditto; if we're using `@ts-ignore` anyways I'd advocate directly referencing `instance.iframeRef.current`.